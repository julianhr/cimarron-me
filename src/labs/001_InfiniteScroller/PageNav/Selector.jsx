import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'


const Root = styled.div`
  padding: 20px 0 40px;
  display: flex;
  flex-direction: column;

  label {
    display: block;
    padding-bottom: 5px;
  }

  select {
    background-color: ${({ theme }) => theme.colors.field.background};
  }
`

function Selector({ label, keys, titles, selected, onChange }) {
  const renderOptions = () => {
    return keys.map((key, i) => (
      <option
        key={key}
        value={key.toString()}
      >
        {titles[i]}
      </option>
    ))
  }

  return (
    <Root>
      <label htmlFor='scroller-type'>{label}</label>
      <select
        id='scroller-type'
        value={selected}
        onChange={(event) => onChange(event.target.value)}
      >
        {renderOptions()}
      </select>
    </Root>
  )
}

Selector.propTypes = {
  label: PropTypes.string.isRequired,
  keys: PropTypes.array.isRequired,
  titles: PropTypes.array.isRequired,
  selected: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  onChange: PropTypes.func.isRequired,
}

export default Selector
