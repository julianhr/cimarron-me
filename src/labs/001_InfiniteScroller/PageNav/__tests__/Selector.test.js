import React from 'react'
import { render, mount, shallow } from 'enzyme'
import { create } from 'react-test-renderer'

import Selector from '../Selector'
import RootMock from '../../__tests__/__mocks__/RootMock'


describe('Selector', () => {
  const testProps = {
    label: 'test label',
    keys: 'a b c'.split(' '),
    titles: 'A B C'.split(' '),
    selected:  'B',
    onChange:  () => 'test onChange',
  }

  const renderer = (props) => mount(<RootMock><Selector {...props} /></RootMock>)
  let wrapper

  afterEach(() => {
    if (wrapper) { wrapper.unmount() }
  })

  describe('props', () => {
    test('label is required', () => {
      const props = {...testProps}
      delete props.label
      expect(() => renderer(props)).toThrow()
    })

    test('keys is required', () => {
      const props = {...testProps}
      delete props.keys
      expect(() => renderer(props)).toThrow()
    })

    test('titles is required', () => {
      const props = {...testProps}
      delete props.titles
      expect(() => renderer(props)).toThrow()
    })

    test('selected is not required', () => {
      const props = {...testProps}
      delete props.selected
      expect(() => renderer(props)).not.toThrow()
    })

    test('onChange is required', () => {
      const props = {...testProps}
      delete props.onChange
      expect(() => renderer(props)).toThrow()
    })
  })

  describe('snapshots', () => {
    it('with all props', () => {
      wrapper = renderer(testProps)
      expect(wrapper).toMatchSnapshot()
    })

    it('without selected prop', () => {
      const props = {...testProps}
      delete props.selected
      wrapper = renderer(props)
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('integration', () => {
    it('calls #onChange when `select` changes ', () => {
      const mockOnChange = jest.spyOn(testProps, 'onChange')
      wrapper = renderer(testProps)
      wrapper.find('select').simulate('change', '')
      expect(mockOnChange).toHaveBeenCalled()
    })
  })
})
