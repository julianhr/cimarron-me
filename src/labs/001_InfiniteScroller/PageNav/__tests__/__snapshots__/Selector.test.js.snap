// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selector snapshots with all props 1`] = `
Array [
  <ForwardRef(render)
    styles={
      Object {
        "map": undefined,
        "name": "1e2xhh9-globalStyles-globalStyles",
        "next": undefined,
        "styles": "body{color:#3D3D3D,lineHeight 1.3em;}p,h1,h2,h3,h4,h5,h6,figure{margin:0;}h1,h2,h3,h4,h5,h6{color:#AB164D;}p{padding:.8em 0;}h1{font-size:44;}h2{line-height:2em;font-size:34;font-weight:500;}h3{font-size:24;font-weight:600;}h4{line-height:1.3em;font-size:18;}h5{font-size:12;}h6{font-size:8;}code{color:#D90000;font-size:14;}label:globalStyles;label:globalStyles;",
      }
    }
  />,
  <Root>
    <div
      className="css-1jt9zxq-Root el5yv3l0"
    >
      <label
        htmlFor="scroller-type"
      >
        test label
      </label>
      <select
        id="scroller-type"
        onChange={[Function]}
        value="B"
      >
        <option
          key="a"
          value="a"
        >
          A
        </option>
        <option
          key="b"
          value="b"
        >
          B
        </option>
        <option
          key="c"
          value="c"
        >
          C
        </option>
      </select>
    </div>
  </Root>,
]
`;

exports[`Selector snapshots without selected prop 1`] = `
Array [
  <ForwardRef(render)
    styles={
      Object {
        "map": undefined,
        "name": "1e2xhh9-globalStyles-globalStyles",
        "next": undefined,
        "styles": "body{color:#3D3D3D,lineHeight 1.3em;}p,h1,h2,h3,h4,h5,h6,figure{margin:0;}h1,h2,h3,h4,h5,h6{color:#AB164D;}p{padding:.8em 0;}h1{font-size:44;}h2{line-height:2em;font-size:34;font-weight:500;}h3{font-size:24;font-weight:600;}h4{line-height:1.3em;font-size:18;}h5{font-size:12;}h6{font-size:8;}code{color:#D90000;font-size:14;}label:globalStyles;label:globalStyles;",
      }
    }
  />,
  <Root>
    <div
      className="css-1jt9zxq-Root el5yv3l0"
    >
      <label
        htmlFor="scroller-type"
      >
        test label
      </label>
      <select
        id="scroller-type"
        onChange={[Function]}
      >
        <option
          key="a"
          value="a"
        >
          A
        </option>
        <option
          key="b"
          value="b"
        >
          B
        </option>
        <option
          key="c"
          value="c"
        >
          C
        </option>
      </select>
    </div>
  </Root>,
]
`;
